csrc/scheduler/normalization_inner_outer_tma_ws.cpp

```diff
@@ -31,210 +31,199 @@ void getHeuristics(
+        // other smems are stacked on top of it directly, see
+        // assignNextAddress in StackBasedSharedMemAllocator
+        int64_t reduction_workspace_size = roundUpToMultiple(
+            iter_unroll * bdimx * bdimy * tmp_gmem_dtype_size, 128);
+        return (int64_t)dev_prop->sharedMemPerBlockOptin >=
+            buffer_size + mbarrier_size + reduction_workspace_size;
+      };
+
+  // Check register usage,  it is calculated as:
+  // (1) Used to further cache circular buffered tv, optional [true]
+  // (2) Used to further cache non-circular buffered tv, optional [true]
+  // (3) Used to cache partial outer reduction results
+  // (4) overhead for indexing, etc.
+  const bool is_circular_buffer_regs_cached = true;
+  const bool is_non_circular_buffer_regs_cached = true;
+  // Given a smem buffer size, calculate the number of registers pre thread
+  // required to cache it in registers. The total required register size may be
+  // larger than smem size due to non-divisible split.
+  auto smem_to_regs =
+      [&](int64_t smem_buffer_size, int64_t bdimx, int64_t iter_unroll) {
+        int persistent_batch = ceilDiv(after_vect, bdimx);
+        int buffer_per_element = smem_buffer_size / inner_dim_numel;
+        int elements_per_thread = persistent_batch * iter_unroll * vect_factor;
+        int buffer_per_thread = buffer_per_element * elements_per_thread;
+        return buffer_per_thread / scheduler_utils::bytes_per_register;
+      };
+  auto is_enough_regs = [&](int64_t iter_unroll, int64_t bdimx) {
+    int64_t reg_count = 0;
+    // cache circular buffered tv
+    if (is_circular_buffer_regs_cached) {
+      reg_count +=
+          smem_to_regs(circular_buffered_smem_size, bdimx, iter_unroll);
     }
-    std::string toString() const {
-      std::stringstream ss;
-      ss << "inner_vect: " << inner_vect << ", inner_batch: " << inner_batch
-         << ", gdimy: " << gdimy
-         << ", tmp_gmem_write_vect: " << tmp_gmem_write_vect
-         << ", vectorization_factor_outer: " << vectorization_factor_outer
-         << ", bdimx: " << bdimx << ", warps_per_sm: " << warps_per_sm
-         << ", required_register_per_thread: " << required_register_per_thread
-         << ", available_register_per_thread: "
-         << available_register_per_thread;
-      return ss.str();
+
+    // cache non-circular buffered tv
+    if (is_non_circular_buffer_regs_cached) {
+      reg_count +=
+          smem_to_regs(non_circular_buffered_smem_size, bdimx, iter_unroll);
     }
+    // regs for partial outer reduction results.
+    reg_count += regs_buffer_size / scheduler_utils::bytes_per_register;
+    // regs for indexing, etc.
+    reg_count += scheduler_utils::register_overhead;
+    // total usage should be less than 255
+    return reg_count <= scheduler_utils::max_registers_per_thread;
   };
+  // bdimx: number of threads for inner dim.
+  int64_t bdimx = std::max(128, hp_threads_per_block_min);
```

Comment:
```suggestion
  int64_t bdimx = std::max(int64_t(128), hp_threads_per_block_min);
```
**[Suggested by AI]** The changes made to the nvfuser project include:

1. **Code Update in `normalization_inner_outer_tma_ws.cpp`:**
   - The modification changes the way the variable `bdimx` is defined. It updates the `std::max` function call to ensure that the type of the constant `128` is explicitly set to `int64_t`, enhancing type safety: 
     ```cpp
     int64_t bdimx = std::max(int64_t(128), hp_threads_per_block_min);
     ```

2. **Python Script Update in `utils.py`:**
   - Two new command-line argument options for a parser are added:
     - `--debinfo`: This argument allows the user to enable additional debug information when running the script.
     - `--debinfo-mode`: This argument is intended for building `nvfuser` in release mode with debug information, improving the detail available during debugging.

Together, these changes aim to enhance the robustness of the C++ code and improve user options for debugging through the Python utility script.